[DirectX] Add `extract-section` to `llvm-objcopy` and implement it for `DXContainer`

inbelic opened the pull request:
This pr adds the `extract-section` option to `llvm-objcopy` as a common option. It differs from `dump-section` as it will produce a standalone object with just one section, as opposed to just the section contents.

For more context as to other options considered, see https://github.com/llvm/llvm-project/pull/153265#issuecomment-3195696003.

This difference in behaviour is used for DXC compatibility with `extract-rootsignature` and `/Frs`.

This pr then implements this functionality for `DXContainer` objects.

This is the second step of https://github.com/llvm/llvm-project/issues/150277 to implement as a compiler action that invokes `llvm-objcopy` for functionality.

This also completes the implementation of `extract-rootsignature` as described in https://github.com/llvm/llvm-project/issues/149560.

inbelic committed:
add support for `extract-section` for `dxcontainer`

inbelic committed:
add basic test cases

inbelic committed:
add failure testcase

inbelic committed:
add common option `extract-section`

inbelic committed:
comment touch up

inbelic committed:
actually enable option for DXContainer

inbelic committed:
self-review: no need to alloc

inbelic committed:
clang-format

jh7370 reviewed (commented):
Please update the docs with the new option.

Please note that I am off for much of next week and may not have time to review this further for a week or two.

inbelic committed:
review: alphabetical order

inbelic committed:
review: add ELF block to denote unhandled option

inbelic committed:
review: add option docs

inbelic committed:
review: rename tests

inbelic committed:
self-review: misc

- update static sampler offset now that it is computed correctly
- update test comment

inbelic commented on llvm/test/tools/llvm-objcopy/DXContainer/extract-basic.test:
```diff
@@ -0,0 +1,306 @@
+## Tests that a separate DXContainer is created for the RTS0 (root signature)
```
Renamed the files.
I did note that other similar testing had a case for single vs many option(s) specified, so they are left as-is for now. But happy to condense them

jh7370 commented on llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp:
```diff
@@ -28,6 +52,15 @@ static Error handleArgs(const CommonConfig &Config, Object &Obj) {
       return Config.ToRemove.matches(P.Name);
     };
 
+  for (StringRef Flag : Config.ExtractSection) {
```
Since this has no impact on the section removal stuff, I'd move it to the start of the function.

jh7370 reviewed (commented):
Please don't force push to branches under active review as it impairs the reviewing experience (and consequenetly is against LLVM policy). If rebasing is needed, do a merge from the branch you want to update to. The merge commit will disappear when you Squash & Merge).

inbelic committed:
review: fix up comments

inbelic committed:
review: improve testing readability and strictness

inbelic committed:
review: hande extract before any removal

inbelic committed:
review: add test for invalid format err

inbelic commented on llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp:
```diff
@@ -11,13 +11,37 @@
+      auto Write = [&OutFilename, &PartObj](raw_ostream &Out) -> Error {
+        DXContainerWriter Writer(PartObj, Out);
+        if (Error E = Writer.write())
+          return createFileError(OutFilename, std::move(E));
```
At the moment, the underlying `Writer.write()` can not return an error but has the error return type to match the interface of the other object writers.

bogner commented on llvm/test/tools/llvm-objcopy/DXContainer/extract-section-basic.test:
```diff
@@ -0,0 +1,320 @@
+# CHECK-NEXT:   PartCount:     1
+# CHECK-NEXT:   PartOffsets:   [ 36 ]
+# CHECK-NEXT:   Parts:
+# CHECK-NOT: DXIL
+# CHECK-NOT: SFI0
+# CHECK-NOT: HASH
+# CHECK-NOT: ISG1
+# CHECK-NOT: OSG1
```
Technically these sections aren't actually ordered, though in practice I guess they'll always happen to be in this order so it's probably fine. Could we do something like `CHECK-NOT: - Name: {{DXIL|SFI0|HASH|ISG1|OSG1|PSV0}}` to be more robust against ordering though? We could then repeat this same check-not after the RTS0 checks.

bogner commented on llvm/test/tools/llvm-objcopy/DXContainer/extract-section-basic.test:
```diff
@@ -0,0 +1,320 @@
+      Size:            431
+      DXILMajorVersion: 1
+      DXILMinorVersion: 7
+      DXILSize:        1700
+      DXIL:            [ 0x42, 0x43, 0xC0, 0xDE, 0x21, 0xC, 0x0, 0x0, 0xA6,
```
Do we really need this entire DXIL section to test this?

bogner commented on llvm/test/tools/llvm-objcopy/DXContainer/extract-section-headers.test:
```diff
@@ -0,0 +1,59 @@
+# CHECK-NEXT: PartOffsets:   [ 36 ]
+# CHECK-NEXT: Parts:
+
+# CHECK-NOT: FKE0
+# FKE1: Name:            FKE1
+# FKE4-NOT: FKE1
+# CHECK-NOT: FKE2
+# CHECK-NOT: FKE3
+# FKE1-NOT: FKE4
+# FKE4: Name:            FKE4
+# CHECK-NOT: FKE5
+# CHECK-NOT: DXIL
```
Lining things up here would go a long way for readability

```suggestion
# CHECK-NOT: FKE0
# FKE1:      Name: FKE1
# FKE4-NOT:  FKE1
# CHECK-NOT: FKE2
# CHECK-NOT: FKE3
# FKE1-NOT:  FKE4
# FKE4:      Name: FKE4
# CHECK-NOT: FKE5
# CHECK-NOT: DXIL
```

bogner reviewed (commented):
LGTM - I'll leave it to @jh7370 to give the approval though.

inbelic committed:
review: improve test readability and flexibility

jh7370 reviewed (commented):
Please don't resolve comment threads I've started - I use the resolution status to determine whether I've decided I'm happy with the resolution.

inbelic committed:
review: clean up test cases

- makes key value spacing uniform and minimal of yaml objects
- simplify the tests by using --implicit-check-not and CHECK-NEXT for
the valid parts

jh7370 reviewed (approved):
LGTM, thanks!

inbelic committed:
Merge branch 'main' into inbelic/dx-objcopy-extract
